File: prime.cpp
#include <iostream>
#include <vector>

bool is_prime(int num, const std::vector<int>& primes) {
    for (int index = 0; index < primes.size(); ++index) {
        int i = primes[index];
        if (num % i == 0) {
            return false;
        }
    }
    return true;
}

void notDisasterCode() {
    std::vector<int> unique_primes;

    for (int i = 2; i < 2500; ++i) {
        int curr = i;
        if (is_prime(curr, unique_primes)) {
            unique_primes.push_back(curr);
        } else {
            for (int index = 0; index < unique_primes.size(); ++index) {
                int j = unique_primes[index];
                if (curr % j == 0) {
                    unique_primes.push_back(j);
                    while (curr % j == 0) {
                        curr /= j;
                    }
                }
                if (curr == 1) {
                    break;
                }
            }
        }
    }
}

int main() {
    double time = 0;
    for (int i = 0; i < 5; ++i) {
        auto start = std::chrono::high_resolution_clock::now();
        notDisasterCode();
        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double> duration = end - start;
        time += std::chrono::duration<double>(duration).count();
    }
    time = time / 5;
    std::cout << "Average execution time after 5 runs: " << time << " seconds" << std::endl;

    return 0;
}

File: prime.py
import timeit


def disasterCode():
    unique_primes = []
    for i in range (2,2500):
        curr = i
        if is_prime(curr, unique_primes):
            unique_primes.append(curr)
        else:
            for j in unique_primes:
                if curr % j == 0:
                    unique_primes.append(j)
                    while curr % j == 0:
                        curr /= j
                if curr == 1:
                    break


def is_prime(num, primes):
    for i in primes:
        if num % i == 0:
            return False
    return True

# Benchmark the code


if __name__ == "__main__":
    benchmark_code = "disasterCode()"
    setup_code = "from __main__ import disasterCode"

    # Measure the execution time of disasterCode function
    times = []
    for i in range(0,5):
        times.append(timeit.timeit(benchmark_code, setup=setup_code, number=1))

    res = sum(times)/5

    print(f"Average execution time after 5 runs: {res:.6f} seconds")
